telegram/ext/_application.py

```diff
@@ -0,0 +1,1212 @@
+        '_chat_ids_to_be_updated_in_persistence',
+        '_concurrent_updates',
+        '_concurrent_updates_sem',
+        '_conversation_handler_conversations',
+        '_initialized',
+        '_running',
+        '_user_data',
+        '_user_ids_to_be_deleted_in_persistence',
+        '_user_ids_to_be_updated_in_persistence',
+        'bot',
+        'bot_data',
+        'chat_data',
+        'context_types',
+        'error_handlers',
+        'handlers',
+        'job_queue',
+        'persistence',
+        'update_queue',
+        'updater',
+        'user_data',
+    )
+
+    def __init__(
+        self: 'Application[BT, CCT, UD, CD, BD, JQ]',
+        *,
+        bot: BT,
+        update_queue: asyncio.Queue,
+        updater: Optional[Updater],
+        job_queue: JQ,
+        concurrent_updates: Union[bool, int],
+        persistence: Optional[BasePersistence],
+        context_types: ContextTypes[CCT, UD, CD, BD],
+    ):
+        if not was_called_by(
+            inspect.currentframe(), Path(__file__).parent.resolve() / '_applicationbuilder.py'
+        ):
+            warn(
+                '`Application` instances should be built via the `ApplicationBuilder`.',
+                stacklevel=2,
+            )
+
+        self.bot = bot
+        self.update_queue = update_queue
+        self.job_queue = job_queue
+        self.context_types = context_types
+        self.updater = updater
+        self.handlers: Dict[int, List[Handler]] = {}
+        self.error_handlers: Dict[Callable, Union[bool, DefaultValue]] = {}
+
+        if isinstance(concurrent_updates, int) and concurrent_updates < 0:
+            raise ValueError('`concurrent_updates` must be a non-negative integer!')
+        if concurrent_updates is True:
+            concurrent_updates = 4096
+        self._concurrent_updates_sem = asyncio.BoundedSemaphore(concurrent_updates or 1)
+        self._concurrent_updates: int = concurrent_updates or 0
+
+        self.bot_data = self.context_types.bot_data()
+        self._user_data: DefaultDict[int, UD] = defaultdict(self.context_types.user_data)
+        self._chat_data: DefaultDict[int, CD] = defaultdict(self.context_types.chat_data)
+        # Read only mapping
+        self.user_data: Mapping[int, UD] = MappingProxyType(self._user_data)
+        self.chat_data: Mapping[int, CD] = MappingProxyType(self._chat_data)
+
+        self.persistence: Optional[BasePersistence] = None
+        if persistence and not isinstance(persistence, BasePersistence):
+            raise TypeError("persistence must be based on telegram.ext.BasePersistence")
+        self.persistence = persistence
+
+        # Some book keeping for persistence logic
+        self._chat_ids_to_be_updated_in_persistence: Set[int] = set()
+        self._user_ids_to_be_updated_in_persistence: Set[int] = set()
+        self._chat_ids_to_be_deleted_in_persistence: Set[int] = set()
+        self._user_ids_to_be_deleted_in_persistence: Set[int] = set()
```

Comment:
I kinda forgot why we're doing this.. Is it because updating the respective `*_data` for the chat on every update is excessive?

Reply:
yes. Currently each `process_update` call *plus* each `run_async` call calls `update_persistence`, which can become quite a number of calls …